### PR TITLE
modified task status flow 

### PR DESCRIPTION
**Pull request by skv93-coder**
Date: 12-Mar-24

Developer Name: @skv93-coder 

---

## Issue Ticket Number

#554 

## Description
This PR adds message when a user change his task status according to the issue #554. and also removes the functionality marking task done on making progress 100% 

### Documentation Updated?

- [ ] Yes
- [X] No

### Under Feature Flag

- [X] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
[screen-capture (10).webm](https://github.com/Real-Dev-Squad/website-my/assets/68867418/e9ead73c-007e-4dc1-a626-5af612b69346)

[screen-capture (11).webm](https://github.com/Real-Dev-Squad/website-my/assets/68867418/c4d37cfd-7b04-42f1-a8d0-d813f1b434fb)
<details>
<summary>Screenshot 1</summary>

</details>

## Test Coverage
Screenshot from 2024-03-22 00-24-52

<details>
<summary>Screenshot 1</summary>

</details>

## Additional Notes
`resetCurrentTask` 
variable simply holds a function passed to the action 
`handleUpdateTask`
which on call will reset the both task status and progress bar.
earlier it was named `error` and passed imperatively.

### CODE REVIEW

File: app/controllers/tasks.js [hunk 1]
```diff
@@ -263,15 +277,71 @@ export default class TasksController extends Controller {
...
+      currentTask.status === this.TASK_KEYS.BLOCKED;
+    const isNewTaskStatusInProgress =
+      taskData.status === this.TASK_KEYS.IN_PROGRESS;
+    const isNewTaskStatusBlock = taskData.status === this.TASK_KEYS.BLOCKED;
+    const isCurrProgress100 =
+      parseInt(currentTask.percentCompleted || 0) === 100;
+    const isCurrProgress0 = parseInt(currentTask.percentCompleted || 0) === 0;
+
+    if (this.dev && taskData.status) {
+      if (
```
Comment (Atifsid) [inline app/controllers/tasks.js:296]: Can't we somehow use switch case?
I'm seeing too many if statements

File: app/controllers/tasks.js [hunk 2]
```diff
@@ -263,15 +277,71 @@ export default class TasksController extends Controller {
...
+        isCurrentTaskStatusBlock &&
+        !isNewTaskStatusInProgress &&
+        !isCurrProgress100
+      ) {
+        this.message = `The progress of current task is ${currentTask.percentCompleted}%. Proceeding further will make task progress 100%.`;
+        this.showModal = true;
+        this.buttonRequired = true;
+        this.tempTaskId = taskId;
+        this.taskFields.percentCompleted = 100;
```
Comment (Atifsid) [inline app/controllers/tasks.js:330]: `this.showModal, this.buttonRequired, this.tempTaskId`
Maybe you can move these out as these are common for all the conditions

File: app/controllers/tasks.js [hunk 3]
```diff
@@ -263,15 +277,71 @@ export default class TasksController extends Controller {
...
+        return;
+      }
+      if (
+        isCurrentTaskStatusBlock &&
+        !isNewTaskStatusInProgress &&
+        !isCurrProgress100
```
Comment (Atifsid) [inline app/controllers/tasks.js:324]: Move !isCurrProgress100 to the parent if, I see it is false for all the conditions

File: app/controllers/tasks.js [hunk 4]
```diff
@@ -97,16 +97,23 @@ export default class TasksController extends Controller {
...
+      ) {
+        requestBody.status = 'COMPLETED';
       }
       requestBody.percentCompleted = parseInt(taskCompletionPercentage);
     }
     return requestBody;
   }
-
+  getTaskById(taskId) {
+    const indexOfSelectedTask = this.allTasks.findIndex(
```
Comment (Atifsid) [inline app/controllers/tasks.js:111]: what if indexOfSelectedTask === -1 ?

File: app/controllers/tasks.js (hunk 1 above)
Comment (iamitprakash) [reply app/controllers/tasks.js:296]: > Can't we somehow use switch case? I'm seeing too many if statements If we can't then minimize the if conditions as suggested in other comments.

good suggestion

File: app/controllers/tasks.js (hunk 3 above)
Comment (skv93-coder) [reply app/controllers/tasks.js]: Done

File: app/controllers/tasks.js (hunk 2 above)
Comment (skv93-coder) [reply app/controllers/tasks.js]: Done

File: app/controllers/tasks.js (hunk 1 above)
Comment (skv93-coder) [reply app/controllers/tasks.js:320]: Done,

File: app/controllers/tasks.js (hunk 4 above)
Comment (skv93-coder) [reply app/controllers/tasks.js:111]: It can not be -1 as that means there is task id which was not returned by the backend.(User is not involved in creating id)

File: app/controllers/tasks.js [hunk 5]
```diff
@@ -262,16 +276,70 @@ export default class TasksController extends Controller {
...
+        switch (currentTask.status) {
+          case this.TASK_KEYS.IN_PROGRESS:
+            if (!isNewTaskStatusBlock) {
+              this.showTaskChangeInfoModal(
+                TASK_MESSAGES.CHANGE_TO_100_PROGRESS
+              );
+              this.taskFields.percentCompleted = 100;
+              return;
+            }
+            break;
```
Comment (Atifsid) [inline app/controllers/tasks.js:309]: my suggestion was to move those variables out of the if statements, not to create a separate function but it looks good
you can move this.taskFields.percentCompleted = 100; out of the switch case and remove unnecessary break statements as the case will end as soon as you return

File: app/controllers/tasks.js [hunk 6]
```diff
@@ -262,16 +276,70 @@ export default class TasksController extends Controller {
...
+          case this.TASK_KEYS.BLOCKED:
+            if (!isNewTaskStatusInProgress) {
+              this.showTaskChangeInfoModal(
+                `The progress of current task is ${currentTask.percentCompleted}%. ${TASK_MESSAGES.CHANGE_TO_100_PROGRESS}`
+              );
+              this.taskFields.percentCompleted = 100;
+              return;
+            }
+            break;
+          default:
```
Comment (Atifsid) [inline app/controllers/tasks.js:319]: why do we need a empty default statement

Comment (skv93-coder) [reply app/controllers/tasks.js:319]: Removed

File: tests/integration/components/tasks-test.js
Comment (Atifsid) [inline tests/integration/components/tasks-test.js:1]: Get this reviewed properly

**Review by Atifsid — DISMISSED** [review]

**Review by palakgupta2712 — DISMISSED** [review]

File: app/controllers/tasks.js [hunk 7]
```diff
@@ -262,16 +276,62 @@ export default class TasksController extends Controller {
...
+    const isCurrentTaskStatusBlock =
+      currentTask.status === this.TASK_KEYS.BLOCKED;
+    const isCurrentTaskStatusInProgress =
+      this.TASK_KEYS.IN_PROGRESS === currentTask.status;
+    const isNewTaskStatusInProgress =
+      taskData.status === this.TASK_KEYS.IN_PROGRESS;
+    const isNewTaskStatusBlock = taskData.status === this.TASK_KEYS.BLOCKED;
+    const isCurrProgress100 =
+      parseInt(currentTask.percentCompleted || 0) === 100;
+    const isCurrProgress0 = parseInt(currentTask.percentCompleted || 0) === 0;
```
Comment (Achintya-Chatterjee) [inline app/controllers/tasks.js:297]: the logic for determining whether to proceed with the task update or show a modal dialogue can be refactored for better readability and maintainability. Consider breaking down the logic into smaller, more focused functions.

File: app/controllers/tasks.js [hunk 8]
```diff
@@ -262,16 +276,62 @@ export default class TasksController extends Controller {
...
+    if (this.dev && taskData.status) {
+      if (
+        (isCurrentTaskStatusBlock || isCurrentTaskStatusInProgress) &&
+        !isCurrProgress100 &&
+        !isNewTaskStatusBlock &&
+        !isNewTaskStatusInProgress
+      ) {
+        this.taskFields.percentCompleted = 100;
+        this.showTaskChangeInfoModal(
+          `The progress of current task is ${currentTask.percentCompleted}%. ${TASK_MESSAGES.CHANGE_TO_100_PROGRESS}`
```
Comment (Achintya-Chatterjee) [inline app/controllers/tasks.js:309]: The logic for showing the modal dialogue could be consolidated into a separate function to avoid repetition and improve code organization.

File: app/controllers/tasks.js [hunk 9]
```diff
@@ -262,16 +276,62 @@ export default class TasksController extends Controller {
...
+        );
+        return;
+      }
+
+      if (
+        isNewTaskStatusInProgress &&
```
Comment (Achintya-Chatterjee) [inline app/controllers/tasks.js:315]:
```suggestion
        isStatusInProgressChanged &&
```

File: tests/integration/components/tasks-test.js [hunk 10]
```diff
@@ -80,4 +80,79 @@ module('Integration | Component | tasks', function (hooks) {
...
+    assert.equal(ctrl.showModal, false);
+    assert.equal(ctrl.message, '');
+
+    ctrl.allTasks = tasks;
+    ctrl.dev = true;
+    ctrl.taskFields = {
+      status: 'SMOKE_TESTING',
+    };
+
+    // action
```
Comment (Achintya-Chatterjee) [inline tests/integration/components/tasks-test.js:148]: Write unit tests to cover more scenarios, including edge cases and error conditions, to ensure the robustness and reliability of the code.

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by palakgupta2712 — APPROVED** [review]
